tests/Test/Xrefcheck/TooManyRequestsSpec.hs

```diff
@@ -0,0 +1,142 @@
+{- SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+module Test.Xrefcheck.TooManyRequestsSpec where
+
+import Universum
+
+import Control.Concurrent (forkIO, killThread)
+import Control.Exception qualified as E
+import Data.CaseInsensitive qualified as CI
+import Data.Map qualified as M
+import Data.Time.Clock.POSIX
+import Fmt (indentF, pretty, unlinesF)
+import Network.HTTP.Types (Status (..), ok200, serviceUnavailable503, tooManyRequests429)
+import Network.HTTP.Types.Header (hRetryAfter)
+import Test.Hspec
+import Time (sec, (-:-))
+import Web.Firefly (ToResponse (toResponse), route, run)
+
+import Xrefcheck.Config
+import Xrefcheck.Core
+import Xrefcheck.Progress
+import Xrefcheck.Util
+import Xrefcheck.Verify
+
+spec :: Spec
+spec = do
+  describe "429 response tests" $ do
+    it "Returns 200 eventually" $ do
+      let prog = Progress{ pTotal = 1
+                         , pCurrent = 1
+                         , pErrorsUnfixable = 0
+                         , pErrorsFixable = 0
+                         , pTaskTimestamp = Nothing
+                         }
+      checkLinkAndProgressWithServer (mock429 ok200)
+        "http://127.0.0.1:5000/429" prog $ VerifyResult []
+    it "Returns 503 eventually" $ do
+      let prog = Progress{ pTotal = 1
+                         , pCurrent = 1
+                         , pErrorsUnfixable = 1
+                         , pErrorsFixable = 0
+                         , pTaskTimestamp = Nothing
+                         }
+      checkLinkAndProgressWithServer (mock429 serviceUnavailable503)
+        "http://127.0.0.1:5000/429" prog $ VerifyResult
+          [ ExternalHttpResourceUnavailable $
+              Status { statusCode = 503, statusMessage = "Service Unavailable"}
+          ]
+    it "Successfully updates the new retry-after value" $ do
+      E.bracket (forkIO (newIORef 0 >>= mock429 ok200)) killThread $ \_ -> do
```

Comment:
Couple of things. For the purpose of this test:

* the server does not need a `switchRef :: IORef Int` that cycles between 0 and 2. A simpler `callCount` is sufficient.
* the argument passed to `mock429` (`ok200`) will be ignored

Instead of trying to generalize `mock429` to fit the needs of many tests, it's better to write simpler single-user mock servers.

For example, for this test, we can write an inline one-off mock server that does _only_ what's required for the test to pass:


```suggestion
      let 
        -- When called for the first time, returns with a 429 and `Retry-After: 2`.
        -- Subsequent calls will respond with a 200 OK.
        mockServer :: IO ()
        mockServer = do
          callCount <- newIORef @_ @Int 0
          run 5000 $ do
            route "/429retryafterupdate" $ do
              count <- atomicModifyIORef' callCount $ \cc -> (cc + 1, cc)
              pure $
                if count == 0
                then toResponse
                  ( "" :: Text
                  , tooManyRequests429
                  , M.fromList [(CI.map (decodeUtf8 @Text) hRetryAfter, ["2" :: Text])]
                  )
                else toResponse ("" :: Text, ok200)
      E.bracket (forkIO mockServer) killThread $ \_ -> do
```